### ISSUE TITLE
Failure to infer the type when a function is defined later than called from a lambda

### ISSUE THREAD

**Issue by povilasb** [microsoft/pyright#4567]
**Describe the bug**

I am not completely sure if this is a bug or rather undefined behavior.
It is a very corner case issue. Let me consider different scenarios and pyright versions with which type checks fail or pass.

1. Run **pyright@1.1.292** on the provided code snippet. You should see an error
```sh
  /private/tmp/main.py:35:10 - error: Argument of type "(prev: None) -> None" cannot be assigned to parameter "func" of type "(str) -> _Out2T@map" in function "map"
    Type "(prev: None) -> None" cannot be assigned to type "(str) -> _Out2T@map"
      Parameter 1: type "str" cannot be assigned to type "None"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
```

2. The last version I found not to report an error on the same code was `1.1.242`.

3. If I move `_get_date()` declaration up, the checks pass:
```py
def _get_date(d: Data) -> str:
    return "2021-01-01"


f = (
    Flow[Data]()
    .map(lambda prev: _get_date(prev))
    .map(lambda prev: prev)
    .map(lambda _prev: "any value")
)
```

4. Lets say the declaration is still later than it's used, but the initial `Flow[OutT=dict]` (or any built-in type), then the checks pass as well:
```py
f = (
    Flow[dict]()
    .map(lambda prev: _get_date(prev))
    .map(lambda prev: prev)
    .map(lambda _prev: "any value")
)

def _get_date(d: dict) -> str:
    return "2021-01-01"
```

5. The checks pass with the original snippet except with no `.map()` overloads
```py
import typing as t

_OutT = t.TypeVar("_OutT")
_Out2T = t.TypeVar("_Out2T")


class Flow(t.Generic[_OutT]):
    def map(
        self: "Flow[_OutT]",
        func: t.Callable[[_OutT], _Out2T],
        /,
    ) -> "Flow[_Out2T]":
        return t.cast("Flow[_Out2T]", self)


class Data:
    ...


f = (
    Flow[Data]()
    .map(lambda prev: _get_date(prev))
    .map(lambda prev: prev)
    .map(lambda _prev: "any value")
)


def _get_date(d: Data) -> str:
    return "2021-01-01"
```

**Expected behavior**

The provide code snippet to pass the type checks as with **pyright <= 1.1.242**.

**Code**

```py
import typing as t

_OutT = t.TypeVar("_OutT")
_Out2T = t.TypeVar("_Out2T")


class Flow(t.Generic[_OutT]):
    @t.overload
    def map(
        self: "Flow[_OutT]",
        func: t.Callable[[_OutT], Exception],
        /,
    ) -> "Flow[None]":
        ...

    @t.overload
    def map(
        self: "Flow[_OutT]",
        func: t.Callable[[_OutT], _Out2T],
        /,
    ) -> "Flow[_Out2T]":
        ...

    def map(self, obj, /):
        return t.cast("Flow", self)


class Data:
    ...


f = (
    Flow[Data]()
    .map(lambda prev: _get_date(prev))
    .map(lambda prev: prev)
    .map(lambda _prev: "any value")
)


def _get_date(d: Data) -> str:
    return "2021-01-01"
```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.293, which I just published. It will also be included in a future release of pylance.
